src/Microsoft.OData.Core/Metadata/EdmLibraryExtensions.cs

```diff
@@ -1826,23 +1930,33 @@ private static bool ParametersSatisfyFunction(IEdmOperation operation, IList<str
                 parametersToMatch = parametersToMatch.Skip(1);
             }
 
-            List<IEdmOperationParameter> functionParameters = parametersToMatch.ToList();
-
             // if any required parameters are missing, don't consider it a match.
-            if (functionParameters.Where(
-                p => !(p is IEdmOptionalParameter)).Any(
-                    p => parameterNameList.All(
-                        k => !string.Equals(k, p.Name, caseInsensitive ? StringComparison.OrdinalIgnoreCase : StringComparison.Ordinal))))
+            BitArray parameterExists = new BitArray(parameterNameList.Count);
+            foreach (var functionParameter in parametersToMatch)
             {
-                return false;
+                bool matched = false;
+                for (int j = 0; j < parameterNameList.Count; j++)
+                {
+                    if (string.Equals(parameterNameList[j], functionParameter.Name, caseInsensitive ? StringComparison.OrdinalIgnoreCase : StringComparison.Ordinal))
+                    {
+                        matched = true;
+                        parameterExists[j] = true;
+                    }
+                }
+
+                if (!matched && !(functionParameter is IEdmOptionalParameter))
+                {
+                    return false;
+                }
             }
 
             // if any specified parameters don't match, don't consider it a match.
-            if (parameterNameList.Any(
-                k => functionParameters.All(
-                    p => !string.Equals(k, p.Name, caseInsensitive ? StringComparison.OrdinalIgnoreCase : StringComparison.Ordinal))))
+            for (int i = 0; i < parameterNameList.Count; i++)
             {
-                return false;
+                if (!parameterExists[i])

```

Comment:
>parameterExists [](start = 21, length = 15)

Can i do it in O(1)?

Reply:
Nope. 

---
In reply to: [370886886](https://github.com/OData/odata.net/pull/1602#discussion_r370886886) [](ancestors = 370886886)